File: packages/algorithm/Leetcode/hot100/数据结构/二叉树/验证二叉搜索树.ts
import type {TreeNode} from './types'

/* 
二叉搜索树
left < middle < right, 对每一个子树都成立
*/
function isValidBST(root: TreeNode | null): boolean {
    // 遍历过程中维护三个量 node | lower | upper
    const dfs = (node: TreeNode | null, lower: number, upper: number): boolean => {
        if (node === null) {
            return true;
        }

        if (node.val <= lower || node.val >= upper) {
            return false;
        }

        if (!dfs(node.left, lower, node.val)) {
            return false;
        }

        if (!dfs(node.right, node.val, upper)) {
            return false;
        }

        return true;
    }

    return dfs(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};

File: Design/alog/first/数据结构/二叉树/py/验证二叉搜索树.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node: Optional[TreeNode], cur_max: int, cur_min: int) -> bool:
            if not node:
                return True

            if node.val >= cur_max or node.val <= cur_min:
                return False

            return dfs(node.left, node.val, cur_min) and dfs(node.right, cur_max, node.val)

        return dfs(root, int('-inf'), int('inf'))
